lib/Doctrine/ORM/Utility/PersisterHelper.php

```diff
@@ -0,0 +1,141 @@
+<?php
+
+/*
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license. For more information, see
+ * <http://www.doctrine-project.org>.
+ */
+
+namespace Doctrine\ORM\Utility;
+
+use Doctrine\Common\Util\ClassUtils;
+use Doctrine\ORM\EntityManagerInterface;
+use Doctrine\ORM\Mapping\ClassMetadata;
+use Doctrine\ORM\Query\QueryException;
+
+/**
+ * The PersisterHelper contains logic to infer binding types which is used in
+ * several persisters.
+ *
+ * @link   www.doctrine-project.org
+ * @since  2.5
+ * @author Jasper N. Brouwer <jasper@nerdsweide.nl>
+ */
+class PersisterHelper
+{
+    /**
+     * @param mixed                  $value
+     * @param EntityManagerInterface $em
+     *
+     * @return mixed
+     */
+    public static function getValue($value, EntityManagerInterface $em)
+    {
+        if ( ! is_array($value)) {
+            return self::getIndividualValue($value, $em);
+        }
+
+        $newValue = array();
+
+        foreach ($value as $fieldName => $fieldValue) {
+            $newValue[$fieldName] = self::getIndividualValue($fieldValue, $em);
+        }
+
+        return $newValue;
+    }
+
+    /**
+     * @param mixed                  $value
+     * @param EntityManagerInterface $em
+     *
+     * @return mixed
+     */
+    private static function getIndividualValue($value, EntityManagerInterface $em)
+    {
+        if ( ! is_object($value) || ! $em->getMetadataFactory()->hasMetadataFor(ClassUtils::getClass($value))) {
```

Comment:
`hasMetadataFor` is not safe to use - doesn't ensure that the class is loaded from cache, if not used previously.